### DISCUSSION TITLE
Shorter forms of using statements

### DISCUSSION THREAD

**Discussion by a-tk-by** [dotnet/csharplang#2894] [Language Ideas]
C#8 introduces new shorter form of using statement without explicit scope:
```cs
using var file = File.Create(fileName);
SaveFile(file);
```
instead of
```cs
using (var fs = File.Create(fileName))
{
  SaveFile(file);
}
```
but it still requires variable declaration.

Variable declaration in old-style syntax is optional, but in C#8 is required.

Often there is no need to introduce variable but scoped lifetime is still needed, like that:
```cs
SaveFile(using File.Create(fileName));
```

**Comment by DavidArno** [comment]
[Please see this comment](https://github.com/dotnet/csharplang/issues/2235#issuecomment-464395605)

The variable is needed so the compiler can work out the scope of the using declaration through the scope of its variable.

**Comment by DaZombieKiller** [comment]
I'd like to see this expanded to the `unsafe` and `fixed` keywords too, I often find myself wanting to do the same thing with them.

@DavidArno 
> The variable is needed so the compiler can work out the scope of the using declaration through the scope of its variable.

Given that the equivalent code for `Func(using Usable())` would be `using (var u = Usable()) Func(u)`, I'm not seeing how the scope could be ambiguous to the compiler here. Could you elaborate?

EDIT: Unless you're referring to doing something like `var a = using b;`, which I don't think should be covered by this proposal.

**Comment by Richiban** [comment]
If I can attempt to read the mind of the language design team, I don't think this is something they'll go for; it's just shorthand (you can't use this syntax to accomplish anything you couldn't do before), it saves a pretty trivial amount of code when you can use it and you also can only use it in a pretty limited number of circumstances.

Also, it's probably rarely important but it's confusing as to when objects get disposed in nested expressions:

```csharp
var a = new A();

a.M1(x => x.M2(M3(using new C())));
```

When does the `new C()` get disposed? It's pretty confusing, no matter which option you pick.

**Comment by iam3yal** [comment]
> it's just shorthand (you can't use this syntax to accomplish anything you couldn't do before), it saves a pretty trivial amount of code when you can use it and you also can only use it in a pretty limited number of circumstances.

That's not really an argument, plenty of things are just a syntactic sugar and yet they were added to the language for the sake of readability.

In my opinion, the variable is redundant (from the reader/programmer pov)  and reducing the scope of usage is a good thing.

> Also, it's probably rarely important but it's confusing as to when objects get disposed in nested expressions:

Hardly, it's like saying `a.M1(x => x.M2(M3(new C())));` is confusing.

> When does the new C() get disposed? It's pretty confusing, no matter which option you pick.

I beg to differ, it's pretty trivial to understand as long as the programmer is aware to what it lowers to.

Now, I don't know whether a member of the language design team will champion this as it's really niche but I can't see the downsides.

**Comment by Richiban** [comment]
@eyalsk I think I didn't make my point very clearly about why I used that particular piece of code as an example of why it's confusing. I'm saying that there are a number of perfectly valid interpretations of the scope of this disposable value, and even when you know how the feature works it's still difficult to tell.

For example, the code in question: 

```csharp
a.M1(x => x.M2(M3(using new C())));
```

could be interpreted as being equivalent to this (a):

```csharp
using (var $tmp = new C())
{
	a.M1(x => x.M2(M3($tmp)));
}
```

or this (b):

```csharp
a.M1(x => {
	using (var $tmp = new C())
	{
		x.M2(M3($tmp))
	}
});
```

or even this (c):

```csharp
var a = new A();

a.M1(x => {
	var $tmp2;

	using (var $tmp1 = new C())
	{ 
		$tmp2 = M3($tmp1);
	}

	return x.M2($tmp2);
});
```

My point is that it's very difficult to give a quick and easy explanation of what the scope is of that variable. Is it the immediately containing expression? That would be option (c), the most complicated to lower. Is it the immetiatly containing statement-expression? That would be option (b). Or the top-most? That's option (a).

I'm not super against this feature or anything, I'm just saying that it's a whole language feature (and potentially a confusing one at that) all to save a few characters.

**Comment by CyrusNajmabadi** [comment]
FWIW, this seems super wrong to me:

> ```c#
> SaveFile(using File.Create(fileName));
> ```

Why?  Because it looks like the file will be disposed *before* calling into SaveFile.  Really feels oogy.

**Comment by DaZombieKiller** [comment]
@CyrusNajmabadi 
> it looks like the file will be disposed *before* calling into SaveFile.

I hadn't looked at it that way previously. Now I can't unsee it!

**Comment by theunrepentantgeek** [comment]
> That's not really an argument, plenty of things are just a syntactic sugar and yet they were added to the language for the sake of readability.

Yup. 

And they're all things where the benefits have been widespread because they addressed extremely common use cases - such as the `=>` forms for simple methods, properties, and local functions.

I'm sure that you're familiar with the oft quoted [language features start at -100 points](https://blogs.msdn.microsoft.com/ericgu/2004/01/12/minus-100-points/).

**Comment by iam3yal** [comment]
@theunrepentantgeek You're right, I'm just saying that if people want to make a case against a feature, saying that something is a "shorthand" for something else is not good enough; however @CyrusNajmabadi does make a good point against it.

**Comment by Richiban** [comment]
Again, my point was not to downvote the feature merely because it's shorthand; god knows I'm not one to shy away from a shorthand feature!

My point was that it's just shorthand *and* it's potentially confusing *and* it doesn't even save that much anyway.

**Comment by iam3yal** [comment]
@Richiban I wasn't referring to your post specifically in my previous comment but speaking in general, I think I understood your point with the explanation you gave above. :)

**Comment by Richiban** [comment]
@eyalsk Nice, thanks for clearing that up :)

**Comment by ronnygunawan** [comment]
> FWIW, this seems super wrong to me:
> 
> > ```cs
> > SaveFile(using File.Create(fileName));
> > ```
> 
> Why? Because it looks like the file will be disposed _before_ calling into SaveFile. Really feels oogy.

If you put it into context:
```cs
{
    ...
    SaveFile(using File.Create(fileName));
    ...
}
```

You can see the scope of the using statement is within the braces enclosing `SaveFile`. The way I see it, it's the same way as inlining `file` variable in:

```cs
{
    ...
    using var file = File.Create(fileName);
    SaveFile(file);
    ...
}
```

**Comment by iam3yal** [comment]
@ronnygunawan Yeah, that's my thoughts exactly although it does look like it happens before `SaveFile` gets executed so I don't know how I feel about it but from syntactic pov it looks good to me.

**Comment by Legends** [comment]
What about:

`byte[] data = using new BinaryReader(s).ReadBytes((int)s.Length);`

Should be disposed when instance method exits.

**Reply by 333fred** [reply]
Good news, that was in C# 8:

```cs
using byte[] data = new BinaryReader(s).ReadBytes((int)s.Length);
```

https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/using-statement

**Reply by Legends** [reply]
If I would want to dispose a byte array (which is not disposable anyway, then yes), but BR should be disposed here.

**Reply by bernd5** [reply]
How could the compiler see this, would it work for a single temporary disposable?

**Reply by 333fred** [reply]
Oh, I see, you wanted it as a nested expression. I have to agree with Bernd here: I would find the semantics of what exactly is getting disposed and when extremely confusing.

**Reply by Legends** [reply]
@333fred From a user's point of view it is apparent what and when should be disposed.
You mean, from a compiler's point of view, right?! I am not knowledgeable of the innerworkings of a compiler ;-)

**Reply by CyrusNajmabadi** [reply]
It's not apparent to me (as a user).  It's unclear waht the `using` closes over.  How it handles multiple values.  And when it should run.  Just looking at it visually confuses me as to what semantics it woudl have.

**Reply by theunrepentantgeek** [reply]
What's wrong with writing

```
using var br = new BinaryReader(s);
using byte[] data = br.ReadBytes((int)s.Length);
```

Being explicit isn't a bad thing.

**Reply by Legends** [reply]
> It's not apparent to me (as a user). It's unclear waht the `using` closes over. How it handles multiple values. And when it should run. Just looking at it visually confuses me as to what semantics it woudl have.

In this case `using` closes when the instance method exits, but I guess there can be different situations where this would not be so obvious I guess.

@theunrepentantgeek Nothing wrong, but you know, we devs want to cut down typing time when possible ;-)

**Reply by CyrusNajmabadi** [reply]
> In this case using closes when the instance method exits,

How do you handle:

`using new MyDisposable().DoSomethingFluentAndDoNotDispose().DoSomethingNowThatMeansImDone()`

?

I don't get how the language feature would work.  And, as examples like thsi show, as a user i'd be very confused about when this woudl be happening.

**Reply by CyrusNajmabadi** [reply]
> we devs want to cut down typing time when possible ;-)

Code is read several orders of magnitude more than it is written.  Clarity def takes precedence here :)

**Reply by Legends** [reply]
> > In this case using closes when the instance method exits,
> 
> How do you handle:
> 
> `using new MyDisposable().DoSomethingFluentAndDoNotDispose().DoSomethingNowThatMeansImDone()`
> 
> ?
> 
> I don't get how the language feature would work. And, as examples like thsi show, as a user i'd be very confused about when this woudl be happening.

That's one case (fluent) I thought of, that could be troublesome.
In this case it's clear, in others it might not.

**Reply by CyrusNajmabadi** [reply]
> in others it might not.

Then that sounds like something undesirable.  :(

**Reply by theunrepentantgeek** [reply]
> Nothing wrong, but you know, we devs want to cut down typing time when possible ;

Frankly, I'm more interested in reducing how long it takes to comprehend my code than how long it takes to type.

Typing speed hasn't been my bottleneck when coding for a very long time.

**Reply by HaloFour** [reply]
See: #451

It suggested a syntax like:

```cs
byte[] data = using (var br = new BinaryReader(s)) br.ReadBytes((int)s.Length);
```

I do believe that this would be equivalent to the following, though:

```cs
byte[] data;
using (var br = new BinaryReader(s)) {
    data = dr.ReadBytes((int)s.Length);
}
```
